Screens API refactor

maxkatz6 opened the pull request:
## What does the pull request do?

1. Screens API now consistently returns the same instances of Screen class for the same physical screen. Previously it would return new instances each time.
2. On physical screen changes, previously retrieved Screen instances now also get updated. No need to re-read all screens each time.
3. New event added that is raised on any screen changed or added (no fine tuning here, sorry, maybe in the future).
4. Adds an API to retrieve screen platform handle, where possible.
5. +DisplayName and +CurrentOrientation properties on screen.
6. Screen is accessible from the TopLevel, not only Window.
7. Adds implementations for all platforms (in this PR - only Windows and MacOS to keep it simpler, Linux is skipped for now)

I skipped Linux because I have no idea how it works and can't properly test it locally.
Which also means, Linux still returns new Screen instance on each read, basically an old behavior.

See https://github.com/AvaloniaUI/Avalonia/issues/15901 for details

Updated API:

```diff
public class WindowBase : TopLevel
{
-    public Screens Screens { get; }
+    public new Screens Screens => base.Screens!;
}
public class TopLevel
{
+    public Screens? Screens { get; }
}

public class Screen
{
+    public string? DisplayName { get; }
+    public ScreenOrientation CurrentOrientation { get; }

+    public IPlatformHandle? TryGetPlatformHandle();
}

public class Screens
{
+    public event EventHandler? Changed;
+    public Screen? ScreenFromTopLevel(TopLevel topLevel);
+    Task<bool> RequestScreenDetails();
}

+public enum ScreenOrientation
+{
+    None,
+    Landscape,
+    Portrait,
+    LandscapeFlipped,
+    PortraitFlipped
+}
```


## Checklist

- [x] Added unit tests (if possible)?
- [x] Added XML documentation to any related classes?
- [x] Consider submitting a PR to https://github.com/AvaloniaUI/avalonia-docs with user documentation

## Breaking changes

`ctor.Screens(IScreenImpl)` is now PrivateAPI. It should not break any app, as IScreenImpl by itself was a private API since 11.0.

## Fixed issues
#15382
#11512
#15637
#15901

maxkatz6 committed:
Draft new API

maxkatz6 committed:
Push reusable ScreensBaseImpl implementation

maxkatz6 committed:
Fix tests and stubs

maxkatz6 committed:
Update ScreensPage sample to work on mobile + show new APIs

maxkatz6 committed:
Reimplement Windows ScreensImpl, reuse existing screens in other places of backend, use Microsoft.Windows.CsWin32 for interop

maxkatz6 committed:
Make X11 project buildable, don't utilize new APIs yet

maxkatz6 committed:
Reimplement macOS Screens API, differenciate screens by CGDirectDisplayID

maxkatz6 committed:
Fix build

maxkatz6 committed:
Adjust breaking changes file (none affect users)

maxkatz6 committed:
Merge branch 'master' into screens-api-refactor

maxkatz6 commented:
One breaking change that I would love to change, but can't - make Bounds and WorkArea nullable, and instead add non-nullable Size. Because we can get `Screen.Size` pretty much on any platform, but screen position and work area are only available on desktop + desktop chromium.
```diff
public class Screen
{
+     public PixelSize Size { get; }
-     public PixelRect Bounds { get; }
+     public PixelRect? Bounds { get; }
-     public PixelRect WorkArea { get; }
+     public PixelRect? WorkArea { get; }
}
```

We might move this specific change to 12.0, cc @kekekeks

maxkatz6 commented on src/Windows/Avalonia.Win32/Avalonia.Win32.csproj:
```diff
@@ -5,8 +5,15 @@
+    <PackageReference Include="System.Runtime.CompilerServices.Unsafe" Version="6.0.0" />
+  </ItemGroup>
+  <ItemGroup>
+    <PackageReference Include="Microsoft.Windows.CsWin32" Version="0.3.106">
```
I also want to use this PR as a first attempt to introduce CsWin32 into the repository.
If this source generator will get annoying enough, we can rewrite it to old-style interop.
I would want to keep it though, if possible.

DJGosnell replied:
I want to use this for the spell checking systems as well. So this is quite helpful to introduce it now.

maxkatz6 commented on native/Avalonia.Native/src/OSX/Screens.mm:
```diff
@@ -1,36 +1,59 @@
+
+    virtual HRESULT GetScreen (
+       CGDirectDisplayID displayId,
+       void** localizedName,
```
Todo: `localizedName` is never assigned a value, forgot about that one.

maxkatz6 commented on native/Avalonia.Native/src/OSX/Screens.mm:
```diff
@@ -43,14 +66,40 @@ virtual HRESULT GetScreen (int index, AvnScreen* ret) override
+private:
+    static void CGDisplayReconfigurationCallBack(CGDirectDisplayID display, CGDisplayChangeSummaryFlags flags, void *userInfo)
+    {
+        auto events = (IAvnScreenEvents*)userInfo;
```
I didn't use any ComPtr here, and just directly passed IAvnScreenEvents* without changing any ownership. Should be fine, I assume, since IAvnScreenEvents owner is C# class that lives whole app execution.

maxkatz6 replied:
Done

maxkatz6 commented on src/Avalonia.Base/Compatibility/UnmanagedCallersOnlyAttribute.cs:
```diff
@@ -0,0 +1,14 @@
+namespace System.Runtime.InteropServices
+{
+    [global::System.AttributeUsage(global::System.AttributeTargets.Method, Inherited = false)]
+    internal sealed class UnmanagedCallersOnlyAttribute : global::System.Attribute
```
FYI: I used function pointers in netstandard2.0, and it even works on net framework.

kekekeks replied:
IIRC one of the reasons for that attribute was to ensure that the function has a proper native->managed transition and is even callable with the corresponding calling convention.

.NET Framework isn't aware of the attribute and isn't likely to setup a proper native->managed transition.

kekekeks replied:
So I believe that it's not safe to do.

maxkatz6 committed:
Fix missing macOS Screen.DisplayName

maxkatz6 committed:
Add more tests + fix screen removal

maxkatz6 committed:
Add screens integration tests

maxkatz6 committed:
Use hash set with comparer when removing screens

maxkatz6 commented on native/Avalonia.Native/src/OSX/Screens.mm:
```diff
@@ -43,14 +66,44 @@ virtual HRESULT GetScreen (int index, AvnScreen* ret) override
             
             ret->Scaling = 1;
```
Does anybody remember why we set scaling to 1? Instead of using an actual value (more likely between 2 and 3 for most main displays).

maxkatz6 replied:
> Our Screen API reports layout scaling

Is it used anywhere like that? It's also not intuitive.
Normally people would use Window.DesktopScaling for that.

maxkatz6 replied:
Other implementations like qtbase and chromium just return dpi scaling, not layout.

maxkatz6 replied:
Will keep it as is for now

maxkatz6 commented:
Added some screen integration tests as well. PR is ready.

maxkatz6 committed:
Merge branch 'master' into screens-api-refactor

kekekeks commented on src/Avalonia.Controls/Platform/Screen.cs:
```diff
@@ -54,12 +104,103 @@ public class Screen
+        {
+            // ReSharper disable NonReadonlyMemberInGetHashCode
+            return _platformHandle?.GetHashCode()
+                   ?? (IsPrimary, DisplayName, CurrentOrientation, Scaling, Bounds, WorkingArea).GetHashCode();
```
I believe that we aren't allowed to change the hash code of a class at runtime. Screen class, however, is now mutable.

maxkatz6 replied:
I will just return `base.GetHashCode()`, when platform handle is not avaialble. This code path is only executed on Linux and for users who manually created Screen object (for some reason that's allowed).

maxkatz6 replied:
Moved these hashcode/equals implementation to another PlatformScreen type.

kekekeks commented on src/Avalonia.Controls/Primitives/PopupPositioning/ManagedPopupPositionerPopupImplHelper.cs:
```diff
@@ -27,12 +27,12 @@ public IReadOnlyList<ManagedPopupPositionerScreenInfo> Screens
             get
             {
-                if (_parent.Screen is null)
+                if (_parent.TryGetFeature<IScreenImpl>() is not { } screenImpl)
```
Why do we need to use a private API here? Is the public API no longer enough?

maxkatz6 replied:
Public API doesn't accept WindowsBaseImpl, it accents WindowsBase. So I would need to use private API either way.

kekekeks commented on src/Avalonia.Native/ScreenImpl.cs:
```diff
@@ -2,66 +2,84 @@
+        public ScreenImpl(Func<IAvnScreenEvents, IAvnScreens> factory)
         {
-            _native = native;
+            _events = new AvnScreenEvents(this);
```
There is no need to keep a reference to the callback class. Just make sure that the native part keeps a proper ComPtr to CCW and do something like:
```cs
using var events = new AvnScreenEvents(this);
_native = factory(events);
```

That way the callback class will be considered to not be referenced from managed code anymore and CCW will get destroyed when the latest reference dies.

Right now the code isn't safe since CCW lifetime is being controlled solely by the managed Dispose.

maxkatz6 replied:
Done

kekekeks commented on src/Windows/Avalonia.Win32/Interop/UnmanagedMethods.cs:
```diff
@@ -1175,10 +1177,8 @@ public struct MOUSEMOVEPOINT
-
-        public delegate bool MonitorEnumDelegate(IntPtr hMonitor, IntPtr hdcMonitor, ref Rect lprcMonitor, IntPtr dwData);
+        public static extern bool EnumDisplayMonitors(nint hdc, nint lprcClip,
+            delegate* unmanaged[Cdecl]<nint, nint, nint, nint, int> lpfnEnum, nint dwData);
```
All WinAPI functions and callbacks are `__stdcall`

maxkatz6 replied:
Replaced this method with CsWinRT generation.
It doesn't use function pointers unfortunately, but that's fine for now.

maxkatz6 commented on native/Avalonia.Native/src/OSX/common.h:
```diff
@@ -89,6 +89,13 @@ template<typename T> class ObjCWrapper {
+@implementation NSScreen (AvNSScreen)
+- (CGDirectDisplayID)av_displayId
+{
+    return [self.deviceDescription[@"NSScreenNumber"] unsignedIntValue];
```
> Also, why do we need our own NSScreen implementation class?

Well, I copy pasted this technique from qtbase code.
https://github.com/qt/qtbase/blob/213127a013b6ecafb36eae1014d5988198736712/src/plugins/platforms/cocoa/qcocoascreen.mm#L797-L804

Although I just found another one a little bit different
https://github.com/qt/qtbase/blob/213127a013b6ecafb36eae1014d5988198736712/src/plugins/platforms/cocoa/qcocoascreen.h#L114-L119

maxkatz6 replied:
> Not sure if we need/want to do something about that.

With current implementation, it will destroy old screen, and create another one. It will be another instance but should be fine in general.
I wonder though, if NSScreen instance by itself is preserved between GPU switches.

maxkatz6 replied:
> Also, why do we need our own NSScreen implementation class? 

Seems to work fine. Can double check though.

maxkatz6 commented on native/Avalonia.Native/src/OSX/Screens.mm:
```diff
@@ -1,36 +1,59 @@
     FORWARD_IUNKNOWN()
+
+    Screens(IAvnScreenEvents* events) {
+        CGDisplayRegisterReconfigurationCallback(CGDisplayReconfigurationCallBack, events);
```
Yeah, I was thinking about that. But not sure if it's needed, since CCW will be alive until managed side holds it, right?

Although now I am thinking that in this corner case, when managed side was destroyed, but CGDisplayRegisterReconfigurationCallback was still executed, this pointer will point into something undefined.

maxkatz6 replied:
Done

maxkatz6 committed:
Make screenimpl safer on macOS as per review

maxkatz6 committed:
Replace UnmanagedCallersOnly usage with source generated EnumDisplayMonitors

maxkatz6 committed:
Remove unused dllimport

maxkatz6 committed:
Only implement GetHashCode and Equals on PlatformScreen subclass, without changing base Screen

maxkatz6 commented:
@kekekeks removed incorrect UnmanagedCallersOnly usage, changed macOS impl to use com pointer, and moved GetHashCode/Equals impl to another class that inherits Screen.

maxkatz6 committed:
Merge branch 'master' into screens-api-refactor
